File: logisim/app/src/main/java/logisim/gui/main/SelectionBase.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.gui.main;

import java.awt.Graphics;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.HashSet;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import logisim.circuit.Circuit;
import logisim.circuit.CircuitMutation;
import logisim.circuit.Wire;
import logisim.comp.Component;
import logisim.comp.ComponentFactory;
import logisim.comp.EndData;
import logisim.data.AttributeSet;
import logisim.data.Bounds;
import logisim.data.Location;
import logisim.gui.main.Selection.Event;
import logisim.gui.main.Selection.Listener;
import logisim.proj.Project;
import logisim.util.CollectionUtil;

class SelectionBase {

	Project proj;
	private ArrayList<Listener> listeners = new ArrayList<>();

	final HashSet<Component> selected = new HashSet<>(); // of selected Components in circuit
	final HashSet<Component> lifted = new HashSet<>(); // of selected Components removed
	final HashSet<Component> suppressHandles = new HashSet<>(); // of Components
	final Set<Component> unionSet = CollectionUtil.createUnmodifiableSetUnion(selected, lifted);

	private Bounds bounds = Bounds.EMPTY_BOUNDS;
	private boolean shouldSnap;

	public SelectionBase(Project proj) {
		this.proj = proj;
	}

	//
	// listener methods
	//
	public void addListener(Listener l) {
		listeners.add(l);
	}

	public void removeListener(Listener l) {
		listeners.remove(l);
	}

	public void fireSelectionChanged() {
		bounds = null;
		computeShouldSnap();
		Event e = new Event(this);
		for (Listener l : listeners) l.selectionChanged(e);
	}

	//
	// query methods
	//
	public Bounds getBounds() {
		if (bounds == null) bounds = computeBounds(unionSet);
		return bounds;
	}

	public Bounds getBounds(Graphics g) {
		Iterator<Component> it = unionSet.iterator();
		if (it.hasNext()) {
			bounds = it.next().getBounds(g);
			while (it.hasNext()) {
				Component comp = it.next();
				Bounds bds = comp.getBounds(g);
				bounds = bounds.add(bds);
			}
		} else bounds = Bounds.EMPTY_BOUNDS;
		return bounds;
	}

	public boolean shouldSnap() {
		return shouldSnap;
	}

	public boolean hasConflictWhenMoved(int dx, int dy) {
		return hasConflictTranslated(unionSet, dx, dy, false);
	}

	//
	// action methods
	//
	public void add(Component comp) {
		if (selected.add(comp)) fireSelectionChanged();
	}

	public void addAll(Collection<? extends Component> comps) {
		if (selected.addAll(comps)) fireSelectionChanged();
	}

	// removes from selection - NOT from circuit
	void remove(CircuitMutation xn, Component comp) {
		boolean removed = selected.remove(comp);
		if (lifted.contains(comp)) if (xn == null) throw new IllegalStateException("cannot remove");
		else {
			lifted.remove(comp);
			removed = true;
			xn.add(comp);
		}

		if (removed) {
			if (shouldSnapComponent(comp))
				computeShouldSnap();
			fireSelectionChanged();
		}
	}

	void dropAll(CircuitMutation xn) {
		if (!lifted.isEmpty()) {
			xn.addAll(lifted);
			selected.addAll(lifted);
			lifted.clear();
		}
	}

	// removes all from selection - NOT from circuit
	void clear(CircuitMutation xn) {
		if (selected.isEmpty() && lifted.isEmpty())
			return;

		if (!lifted.isEmpty()) xn.addAll(lifted);

		selected.clear();
		lifted.clear();
		shouldSnap = false;
		bounds = Bounds.EMPTY_BOUNDS;

		fireSelectionChanged();
	}

	public void setSuppressHandles(Collection<Component> toSuppress) {
		suppressHandles.clear();
		if (toSuppress != null)
			suppressHandles.addAll(toSuppress);
	}

	void duplicateHelper(CircuitMutation xn) {
		HashSet<Component> oldSelected = new HashSet<>(selected);
		oldSelected.addAll(lifted);
		pasteHelper(xn, oldSelected);
	}

	void pasteHelper(CircuitMutation xn, Collection<Component> comps) {
		clear(xn);

		Map<Component, Component> newLifted = copyComponents(comps);
		lifted.addAll(newLifted.values());
		fireSelectionChanged();
	}

	void deleteAllHelper(CircuitMutation xn) {
		for (Component comp : selected) xn.remove(comp);
		selected.clear();
		lifted.clear();
		fireSelectionChanged();
	}

	void translateHelper(CircuitMutation xn, int dx, int dy) {
		Map<Component, Component> selectedAfter = copyComponents(selected, dx, dy);
		for (Entry<Component, Component> entry : selectedAfter.entrySet()) xn.replace(entry.getKey(), entry.getValue());

		Map<Component, Component> liftedAfter = copyComponents(lifted, dx, dy);
		lifted.clear();
		for (Entry<Component, Component> entry : liftedAfter.entrySet()) {
			xn.add(entry.getValue());
			selected.add(entry.getValue());
		}
		fireSelectionChanged();
	}

	//
	// private methods
	//
	private void computeShouldSnap() {
		shouldSnap = false;
		for (Component comp : unionSet)
			if (shouldSnapComponent(comp)) {
				shouldSnap = true;
				return;
			}
	}

	private static boolean shouldSnapComponent(Component comp) {
		Boolean shouldSnapValue = (Boolean) comp.getFactory().getFeature(ComponentFactory.SHOULD_SNAP,
				comp.getAttributeSet());
		return shouldSnapValue == null || shouldSnapValue;
	}

	private boolean hasConflictTranslated(Collection<Component> components, int dx, int dy, boolean selfConflicts) {
		Circuit circuit = proj.getCurrentCircuit();
		if (circuit == null)
			return false;
		for (Component comp : components)
			if (!(comp instanceof Wire)) {
				for (EndData endData : comp.getEnds())
					if (endData != null && endData.isExclusive()) {
						Location endLoc = endData.getLocation().add(dx, dy);
						Component conflict = circuit.getExclusive(endLoc);
						if (conflict != null) if (selfConflicts || !components.contains(conflict))
							return true;
					}
				Location newLoc = comp.getLocation().add(dx, dy);
				Bounds newBounds = comp.getBounds().translate(dx, dy);
				for (Component comp2 : circuit.getAllContaining(newLoc)) {
					Bounds bds = comp2.getBounds();
					if (bds.equals(newBounds)) if (selfConflicts || !components.contains(comp2))
						return true;
				}
			}
		return false;
	}

	private static Bounds computeBounds(Collection<Component> components) {
		if (components.isEmpty()) return Bounds.EMPTY_BOUNDS;
		else {
			Iterator<Component> it = components.iterator();
			Bounds ret = it.next().getBounds();
			while (it.hasNext()) {
				Component comp = it.next();
				Bounds bds = comp.getBounds();
				ret = ret.add(bds);
			}
			return ret;
		}
	}

	private HashMap<Component, Component> copyComponents(Collection<Component> components) {
		// determine translation offset where we can legally place the clipboard
		int dx;
		int dy;
		Bounds bds = computeBounds(components);
		for (int index = 0;; index++) {
			// compute offset to try: We try points along successively larger
			// squares radiating outward from 0,0
			if (index == 0) {
				dx = 0;
				dy = 0;
			} else {
				int side = 1;
				while (side * side <= index)
					side += 2;
				int offs = index - (side - 2) * (side - 2);
				dx = side / 2;
				dy = side / 2;
				// top edge of square
				if (offs < side - 1) dx -= offs;
				else if (offs < 2 * (side - 1)) { // left edge
					offs -= side - 1;
					dx = -dx;
					dy -= offs;
				} else if (offs < 3 * (side - 1)) { // right edge
					offs -= 2 * (side - 1);
					dx = -dx + offs;
					dy = -dy;
				} else {
					offs -= 3 * (side - 1);
					dy = -dy + offs;
				}
				dx *= 10;
				dy *= 10;
			}

			if (bds.getX() + dx >= 0 && bds.getY() + dy >= 0 && !hasConflictTranslated(components, dx, dy, true))
				return copyComponents(components, dx, dy);
		}
	}

	private HashMap<Component, Component> copyComponents(Collection<Component> components, int dx, int dy) {
		HashMap<Component, Component> ret = new HashMap<>();
		for (Component comp : components) {
			Location oldLoc = comp.getLocation();
			AttributeSet attrs = (AttributeSet) comp.getAttributeSet().clone();
			int newX = oldLoc.x() + dx;
			int newY = oldLoc.y() + dy;
			Object snap = comp.getFactory().getFeature(ComponentFactory.SHOULD_SNAP, attrs);
			if (snap == null || (Boolean) snap) {
				newX = Canvas.snapXToGrid(newX);
				newY = Canvas.snapYToGrid(newY);
			}
			Location newLoc = new Location(newX, newY);

			Component copy = comp.getFactory().createComponent(newLoc, attrs);
			ret.put(comp, copy);
		}
		return ret;
	}

	// debugging methods
	public void print() {
		System.err.println(" shouldSnap: " + shouldSnap()); // OK

		boolean hasPrinted = false;
		for (Component comp : selected) {
			System.err.println((hasPrinted ? "         " : " select: ") // OK
					+ comp + "  [" + comp.hashCode() + "]");
			hasPrinted = true;
		}

		hasPrinted = false;
		for (Component comp : lifted) {
			System.err.println((hasPrinted ? "         " : " lifted: ") // OK
					+ comp + "  [" + comp.hashCode() + "]");
			hasPrinted = true;
		}
	}

}

File: LogisimPlus/LogisimPlus/logisim/gui/main/SelectionBase.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using LogisimPlus.Java;
using System;
using System.Collections.Generic;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.gui.main
{

	using Circuit = logisim.circuit.Circuit;
	using CircuitMutation = logisim.circuit.CircuitMutation;
	using Wire = logisim.circuit.Wire;
	using Component = logisim.comp.Component;
	using ComponentFactory = logisim.comp.ComponentFactory;
	using EndData = logisim.comp.EndData;
	using AttributeSet = logisim.data.AttributeSet;
	using Bounds = logisim.data.Bounds;
	using Location = logisim.data.Location;
	using Project = logisim.proj.Project;
	using CollectionUtil = logisim.util.CollectionUtil;

	internal class SelectionBase
	{
		private bool instanceFieldsInitialized = false;

		private void InitializeInstanceFields()
		{
			unionSet = CollectionUtil.createUnmodifiableSetUnion(selected, lifted);
		}

		internal static readonly HashSet<Component> NO_COMPONENTS = Collections.emptySet();

		internal Project proj;
		private List<Selection.Listener> listeners = new List<Selection.Listener>();

		internal readonly HashSet<Component> selected = new HashSet<Component>(); // of selected Components in circuit
		internal readonly HashSet<Component> lifted = new HashSet<Component>(); // of selected Components removed
		internal readonly HashSet<Component> suppressHandles = new HashSet<Component>(); // of Components
		internal HashSet<Component> unionSet;

		private Bounds bounds = Bounds.EMPTY_BOUNDS;
// JAVA TO C# CONVERTER NOTE: Field name conflicts with a method name of the current type:
		private bool shouldSnap_Conflict = false;

		public SelectionBase(Project proj)
		{
			if (!instanceFieldsInitialized)
			{
				InitializeInstanceFields();
				instanceFieldsInitialized = true;
			}
			this.proj = proj;
		}

		//
		// listener methods
		//
		public virtual void addListener(Selection.Listener l)
		{
			listeners.Add(l);
		}

		public virtual void removeListener(Selection.Listener l)
		{
			listeners.Remove(l);
		}

		public virtual void fireSelectionChanged()
		{
			bounds = null;
			computeShouldSnap();
			Selection.Event e = new Selection.Event(this);
			foreach (Selection.Listener l in listeners)
			{
				l.selectionChanged(e);
			}
		}

		//
		// query methods
		//
		public virtual Bounds Bounds
		{
			get
			{
				if (bounds == null)
				{
					bounds = computeBounds(unionSet);
				}
				return bounds;
			}
		}

		public virtual Bounds getBounds(JGraphics g)
		{
			IEnumerator<Component> it = unionSet.GetEnumerator();
// JAVA TO C# CONVERTER TASK: Java iterators are only converted within the context of 'while' and 'for' loops:
			if (it.hasNext())
			{
// JAVA TO C# CONVERTER TASK: Java iterators are only converted within the context of 'while' and 'for' loops:
				bounds = it.next().getBounds(g);
				while (it.MoveNext())
				{
					Component comp = it.Current;
					Bounds bds = comp.getBounds(g);
					bounds = bounds.add(bds);
				}
			}
			else
			{
				bounds = Bounds.EMPTY_BOUNDS;
			}
			return bounds;
		}

		public virtual bool shouldSnap()
		{
			return shouldSnap_Conflict;
		}

		public virtual bool hasConflictWhenMoved(int dx, int dy)
		{
			return hasConflictTranslated(unionSet, dx, dy, false);
		}

		//
		// action methods
		//
		public virtual void add(Component comp)
		{
			if (selected.Add(comp))
			{
				fireSelectionChanged();
			}
		}

		public virtual void addAll<T1>(ICollection<T1> comps) where T1 : logisim.comp.Component
		{
			if (selected.addAll(comps))
			{
				fireSelectionChanged();
			}
		}

		// removes from selection - NOT from circuit
		internal virtual void remove(CircuitMutation xn, Component comp)
		{
			bool removed = selected.Remove(comp);
			if (lifted.Contains(comp))
			{
				if (xn == null)
				{
					throw new System.InvalidOperationException("cannot remove");
				}
				else
				{
					lifted.Remove(comp);
					removed = true;
					xn.add(comp);
				}
			}

			if (removed)
			{
				if (shouldSnapComponent(comp))
				{
					computeShouldSnap();
				}
				fireSelectionChanged();
			}
		}

		internal virtual void dropAll(CircuitMutation xn)
		{
			if (lifted.Count > 0)
			{
				xn.addAll(lifted);
				selected.addAll(lifted);
				lifted.Clear();
			}
		}

		internal virtual void clear(CircuitMutation xn)
		{
			clear(xn, true);
		}

		// removes all from selection - NOT from circuit
		internal virtual void clear(CircuitMutation xn, bool dropLifted)
		{
			if (selected.Count == 0 && lifted.Count == 0)
			{
				return;
			}

			if (dropLifted && lifted.Count > 0)
			{
				xn.addAll(lifted);
			}

			selected.Clear();
			lifted.Clear();
			shouldSnap_Conflict = false;
			bounds = Bounds.EMPTY_BOUNDS;

			fireSelectionChanged();
		}

		public virtual ICollection<Component> SuppressHandles
		{
			set
			{
				suppressHandles.Clear();
				if (value != null)
				{
					suppressHandles.addAll(value);
				}
			}
		}

		internal virtual void duplicateHelper(CircuitMutation xn)
		{
			HashSet<Component> oldSelected = new HashSet<Component>(selected);
			oldSelected.addAll(lifted);
			pasteHelper(xn, oldSelected);
		}

		internal virtual void pasteHelper(CircuitMutation xn, ICollection<Component> comps)
		{
			clear(xn);

			Dictionary<Component, Component> newLifted = copyComponents(comps);
			lifted.addAll(newLifted.Values);
			fireSelectionChanged();
		}

		internal virtual void deleteAllHelper(CircuitMutation xn)
		{
			foreach (Component comp in selected)
			{
				xn.remove(comp);
			}
			selected.Clear();
			lifted.Clear();
			fireSelectionChanged();
		}

		internal virtual void translateHelper(CircuitMutation xn, int dx, int dy)
		{
			Dictionary<Component, Component> selectedAfter = copyComponents(selected, dx, dy);
			foreach (KeyValuePair<Component, Component> entry in selectedAfter.SetOfKeyValuePairs())
			{
				xn.replace(entry.Key, entry.Value);
			}

			Dictionary<Component, Component> liftedAfter = copyComponents(lifted, dx, dy);
			lifted.Clear();
			foreach (KeyValuePair<Component, Component> entry in liftedAfter.SetOfKeyValuePairs())
			{
				xn.add(entry.Value);
				selected.Add(entry.Value);
			}
			fireSelectionChanged();
		}

		//
		// private methods
		//
		private void computeShouldSnap()
		{
			shouldSnap_Conflict = false;
			foreach (Component comp in unionSet)
			{
				if (shouldSnapComponent(comp))
				{
					shouldSnap_Conflict = true;
					return;
				}
			}
		}

		private static bool shouldSnapComponent(Component comp)
		{
			bool? shouldSnapValue = (bool?) comp.Factory.getFeature(ComponentFactory.SHOULD_SNAP, comp.AttributeSet);
			return shouldSnapValue == null ? true : shouldSnapValue.Value;
		}

		private bool hasConflictTranslated(ICollection<Component> components, int dx, int dy, bool selfConflicts)
		{
			Circuit circuit = proj.CurrentCircuit;
			if (circuit == null)
			{
				return false;
			}
			foreach (Component comp in components)
			{
				if (!(comp is Wire))
				{
					foreach (EndData endData in comp.Ends)
					{
						if (endData != null && endData.Exclusive)
						{
							Location endLoc = endData.Location.translate(dx, dy);
							Component conflict = circuit.getExclusive(endLoc);
							if (conflict != null)
							{
								if (selfConflicts || !components.Contains(conflict))
								{
									return true;
								}
							}
						}
					}
					Location newLoc = comp.Location.translate(dx, dy);
					Bounds newBounds = comp.Bounds.translate(dx, dy);
					foreach (Component comp2 in circuit.getAllContaining(newLoc))
					{
						Bounds bds = comp2.Bounds;
						if (bds.Equals(newBounds))
						{
							if (selfConflicts || !components.Contains(comp2))
							{
								return true;
							}
						}
					}
				}
			}
			return false;
		}

		private static Bounds computeBounds(ICollection<Component> components)
		{
			if (components.Count == 0)
			{
				return Bounds.EMPTY_BOUNDS;
			}
			else
			{
				IEnumerator<Component> it = components.GetEnumerator();
// JAVA TO C# CONVERTER TASK: Java iterators are only converted within the context of 'while' and 'for' loops:
				Bounds ret = it.next().getBounds();
				while (it.MoveNext())
				{
					Component comp = it.Current;
					Bounds bds = comp.Bounds;
					ret = ret.add(bds);
				}
				return ret;
			}
		}

		private Dictionary<Component, Component> copyComponents(ICollection<Component> components)
		{
			// determine translation offset where we can legally place the clipboard
			int dx;
			int dy;
			Bounds bds = computeBounds(components);
			for (int index = 0;; index++)
			{
				// compute offset to try: We try points along successively larger
				// squares radiating outward from 0,0
				if (index == 0)
				{
					dx = 0;
					dy = 0;
				}
				else
				{
					int side = 1;
					while (side * side <= index)
					{
						side += 2;
					}
					int offs = index - (side - 2) * (side - 2);
					dx = side / 2;
					dy = side / 2;
					if (offs < side - 1)
					{ // top edge of square
						dx -= offs;
					}
					else if (offs < 2 * (side - 1))
					{ // left edge
						offs -= side - 1;
						dx = -dx;
						dy -= offs;
					}
					else if (offs < 3 * (side - 1))
					{ // right edge
						offs -= 2 * (side - 1);
						dx = -dx + offs;
						dy = -dy;
					}
					else
					{
						offs -= 3 * (side - 1);
						dy = -dy + offs;
					}
					dx *= 10;
					dy *= 10;
				}

				if (bds.X + dx >= 0 && bds.Y + dy >= 0 && !hasConflictTranslated(components, dx, dy, true))
				{
					return copyComponents(components, dx, dy);
				}
			}
		}

		private Dictionary<Component, Component> copyComponents(ICollection<Component> components, int dx, int dy)
		{
			Dictionary<Component, Component> ret = new Dictionary<Component, Component>();
			foreach (Component comp in components)
			{
				Location oldLoc = comp.Location;
				AttributeSet attrs = (AttributeSet) comp.AttributeSet.clone();
				int newX = oldLoc.X + dx;
				int newY = oldLoc.Y + dy;
				object snap = comp.Factory.getFeature(ComponentFactory.SHOULD_SNAP, attrs);
				if (snap == null || ((bool?) snap).Value)
				{
					newX = Canvas.snapXToGrid(newX);
					newY = Canvas.snapYToGrid(newY);
				}
				Location newLoc = new Location(newX, newY);

				Component copy = comp.Factory.createComponent(newLoc, attrs);
				ret[comp] = copy;
			}
			return ret;
		}

		// debugging methods
		public virtual void print()
		{
			Console.Error.WriteLine(" shouldSnap: " + shouldSnap()); // OK

			bool hasPrinted = false;
			foreach (Component comp in selected)
			{
				Console.Error.WriteLine((hasPrinted ? "         " : " select: ") + comp + "  [" + comp.GetHashCode() + "]");
				hasPrinted = true;
			}

			hasPrinted = false;
			foreach (Component comp in lifted)
			{
				Console.Error.WriteLine((hasPrinted ? "         " : " lifted: ") + comp + "  [" + comp.GetHashCode() + "]");
				hasPrinted = true;
			}
		}

	}

}
